src/JITModule.cpp

```diff
@@ -296,9 +281,46 @@ void JITModule::compile_module(std::unique_ptr<llvm::Module> m, const string &fu
     // TODO: If this ever works in LLVM, this would allow profiling of JIT code with symbols with oprofile.
     // listeners.push_back(llvm::createOProfileJITEventListener());
 
-    for (auto &listener : listeners) {
-        ee->RegisterJITEventListener(listener);
+    // Create LLJIT
+    const auto compilerBuilder = [&tm](llvm::orc::JITTargetMachineBuilder /*jtmb*/)
+        -> llvm::Expected<std::unique_ptr<llvm::orc::IRCompileLayer::IRCompiler>> {
+        return std::make_unique<llvm::orc::TMOwningSimpleCompiler>(std::move(*tm));
+    };
+
+    const auto linkerBuilder = [&](llvm::orc::ExecutionSession &session, const llvm::Triple &) {
+        return std::make_unique<llvm::orc::ObjectLinkingLayer>(session);
+    };
+
+    auto JIT = llvm::cantFail(llvm::orc::LLJITBuilder()
+                                  .setDataLayout(target_data_layout)
+                                  .setCompileFunctionCreator(compilerBuilder)
+                                  .setObjectLinkingLayerCreator(linkerBuilder)
+                                  .create());
+
+    auto ctors = llvm::orc::getConstructors(*m);
+    llvm::orc::CtorDtorRunner ctorRunner(JIT->getMainJITDylib());
+    ctorRunner.add(ctors);
+
+    auto dtors = llvm::orc::getDestructors(*m);
+    llvm::orc::CtorDtorRunner *dtorRunner = new llvm::orc::CtorDtorRunner(JIT->getMainJITDylib());
+    dtorRunner->add(dtors);
+
+    // Resolve system symbols (like pthread, dl and others)
+    JIT->getMainJITDylib().addGenerator(llvm::cantFail(llvm::orc::DynamicLibrarySearchGenerator::GetForCurrentProcess(target_data_layout.getGlobalPrefix())));
+
+    // Resolve symbol dependencies.
+    llvm::orc::SymbolMap NewSymbols;
+    auto symbolStringPool = JIT->getExecutionSession().getExecutorProcessControl().getSymbolStringPool();
+    for (const auto &module : dependencies) {
+        for (auto const &iter : module.exports()) {
+            orc::SymbolStringPtr name = symbolStringPool->intern(iter.first);
+            NewSymbols.insert({name, llvm::JITEvaluatedSymbol::fromPointer(iter.second.address)});
+        }
     }
+    llvm::cantFail(JIT->getMainJITDylib().define(llvm::orc::absoluteSymbols(std::move(NewSymbols))));
```

Comment:
`addGenerator` works only for unresolved symbols. However there is a test `correctness_c_function` which replaces JIT extern. So we need redefine every dependency symbol (probably, in reversed order?)